### PR TITLE
[ISSUE #323] Support the expansion of metrics data statistics in RocketMQ-Spring.

### LINKED ISSUE #323: Support the expansion of metrics data statistics in RocketMQ-Spring.
**FEATURE REQUEST**

1. Please describe the feature you are requesting.

Now, in rocketmq-spring project, it dosen't support the expansion of metrics data statistics from the client dimension. This feature can enable:

* Provide standard and diverse metrics representation(e.g. percentile, gauge, histogram).
* Standardization of monitoring API and transport server API, and adapt to open-source monitoring components (e.g. Prometheus).
* Integration with open-source visualization solution(e.g. Grafana).

2. Provide any additional detail on your proposed use case for this feature.

So, we can support this feature in rocketmq-spring.

3. Indicate the importance of this issue to you (blocker, must-have, should-have, nice-to-have). Are you currently using any workarounds to address this issue?

### PR DESCRIPTION
**Pull request by heihaozi**
## What is the purpose of the change

Support the expansion of metrics data statistics in RocketMQ-Spring. This feature can enable:

 * Provide standard and diverse metrics representation(e.g. percentile, gauge, histogram).
 * Standardization of monitoring API and transport server API, and adapt to open-source monitoring components (e.g. Prometheus).
 * Integration with open-source visualization solution(e.g. Grafana).

## Brief changelog

Created an interface called MetricExtension to record the number of messages produced or consumed. By obtaining the implementation of this interface through SPI, developers can easily write their own metrics data statistics method.

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. 
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -176,9 +176,10 @@
                         <exclude>.github/**</exclude>
                         <exclude>src/test/resources/certs/*</exclude>
                         <exclude>src/test/**/*.log</exclude>
-                        <exclude>src/test/resources/META-INF/service/*</exclude>
+                        <exclude>**/src/main/resources/META-INF/services/*</exclude>
```
Comment (vongosling) [inline pom.xml:179]: why did you add prefix placeholder before the root directory /src?

File: rocketmq-spring-boot-samples/rocketmq-consume-demo/src/main/java/org/apache/rocketmq/samples/springboot/AtomicLongMetricExtension.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+public class AtomicLongMetricExtension implements MetricExtension {
+
+    private final Map<String, AtomicLong> producerMessageCountMap = new ConcurrentHashMap<>();
+    private final Map<String, AtomicLong> consumerMessageCountMap = new ConcurrentHashMap<>();
+
+    @Override
+    public void addProducerMessageCount(String topic, int count) {
+        AtomicLong atomicLong = producerMessageCountMap.computeIfAbsent(topic, t -> new AtomicLong());
+        System.out.printf("The count of producer messages for %s is %d.%n", topic, atomicLong.addAndGet(count));
```
Comment (vongosling) [inline rocketmq-spring-boot-samples/rocketmq-consume-demo/src/main/java/org/apache/rocketmq/samples/springboot/AtomicLongMetricExtension.java:35]: It's not best practice to print info in the console. you could make a switch and log it. And do not make too many copies of this class in the different submodule.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.rocketmq.spring.core;
+
+import org.apache.rocketmq.client.consumer.DefaultLitePullConsumer;
+import org.apache.rocketmq.remoting.RPCHook;
+
+public class DefaultLitePullConsumerWithTopic extends DefaultLitePullConsumer {
```
Comment (vongosling) [inline rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java:23]: DefaultLitePullConsumerWithTopic is not a rational name here.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/RocketMQTemplate.java [hunk 4]
```diff
@@ -236,6 +238,7 @@ public void setAsyncSenderExecutor(ExecutorService asyncSenderExecutor) {
             if (delayLevel > 0) {
                 rocketMsg.setDelayTimeLevel(delayLevel);
             }
+            MetricExtensionProvider.addProducerMessageCount(rocketMsg.getTopic(), 1);
```
Comment (vongosling) [inline rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/RocketMQTemplate.java:241]: I suggest you use an atomic data structure instead of plus 1 here:-)

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.rocketmq.spring.metric;
+
+public enum EConsumerMode {
```
Comment (vongosling) [inline rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java:20]: Please do not make an enum here. you could use the existing class.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtension.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+public interface MetricExtension {
+
+    /**
+     * Add current count of message from the producer.
+     *
+     * @param topic topic name
+     * @param count count of message
+     */
+    void addProducerMessageCount(String topic, int count);
```
Comment (vongosling) [inline rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtension.java:28]: int or long should be consistent.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+package org.apache.rocketmq.spring.metric;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+public class MetricExtensionProvider {
```
Comment (vongosling) [inline rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java:27]: Do not make class by yourself. I strongly recommend you use a similar service loader mode in RocketMQ. It has many backoff strategies and verified in the production environment.

File: rocketmq-spring-boot/src/test/java/org/apache/rocketmq/spring/metric/MetricExtensionTest.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    public void addConsumerMessageCount(String topic, int count, EConsumerMode consumerMode) {
+        TOPIC = topic;
+        COUNT = count;
+        CONSUMER_MODE = consumerMode;
+    }
+
+    @Test
+    public void testAddProducerMessageCount() {
+        MetricExtensionProvider.addProducerMessageCount("topic1", 111);
```
Comment (vongosling) [inline rocketmq-spring-boot/src/test/java/org/apache/rocketmq/spring/metric/MetricExtensionTest.java:45]: It could be helpful for your code convincible if you could make more corner cases, such as have you simulated the count of failures?

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java (hunk 3 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java:23]: Compared with DefaultLitePullConsumer, this class just has one more Topic attribute. So, I named this class DefaultLitePullConsumerWithTopic. Do you have any better suggestions for the naming of this class?

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java (hunk 5 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java:20]: I did not find an existing enum. Do you have better suggestions?

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtension.java (hunk 6 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtension.java:28]: The parameter type is always int, and long is not used.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java (hunk 7 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java:27]: The ServiceProvider is in the broker module, RocketMQ-Spring has no dependency on the broker module, so it cannot be used directly.

File: rocketmq-spring-boot/src/test/java/org/apache/rocketmq/spring/metric/MetricExtensionTest.java (hunk 8 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/test/java/org/apache/rocketmq/spring/metric/MetricExtensionTest.java:45]: This unit test is mainly used to test whether the SPI of MetricExtensionProvider is running normally.The scenario where the counting fails you mentioned was considered by the developer when implementing the MetricExtension interface.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java (hunk 7 above)
Comment (vongosling) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java:27]: You could copy and make some improvements. Pls make a reference description when you do it .

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java (hunk 3 above)
Comment (vongosling) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java:23]: Do not make a new class here.

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java (hunk 5 above)
Comment (vongosling) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java:20]: I could not find any use for this enum, if exist, please encapsulate it to the inner class.

File: pom.xml (hunk 1 above)
Comment (heihaozi) [reply pom.xml:179]: I tried it. When running `mvn -B clean apache-rat:check` and building RocketMQ Spring Boot, the rocketmq-spring-boot-samples project will also be checked. If I do not add the prefix placeholder, an unapproved license error will be reported. Like this: [https://travis-ci.org/github/apache/rocketmq-spring/builds/748732956](https://travis-ci.org/github/apache/rocketmq-spring/builds/748732956)

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java (hunk 7 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/MetricExtensionProvider.java:27]: ServiceLoader is a simple Service Provder Framework provided since JDK 1.6. I think the reference description is this:  [https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html](https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html)

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java (hunk 3 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/core/DefaultLitePullConsumerWithTopic.java:23]: Where can this new class be maked? What's your suggestion?

File: rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java (hunk 5 above)
Comment (heihaozi) [reply rocketmq-spring-boot/src/main/java/org/apache/rocketmq/spring/metric/EConsumerMode.java:20]: This is a good idea.

**Comment by vongosling** [comment]
Do we consider the possibility of using aop to finish this feature? @RongtongJin Would you like to help to design here?

**Review by RockteMQ-AI — APPROVED** [review]
## Review: Approved ✅

**PR:** #324 — [ISSUE #323] Support metrics data statistics expansion
**Type:** Enhancement (13 files, +360/-6)

### Assessment
Adds `MetricExtension` SPI for pluggable metrics collection (percentile, gauge, histogram). Uses Java ServiceLoader pattern. Includes test and sample implementations.

### Verdict
✅ Well-designed metrics extension point with proper SPI pattern.

---
*🤖 Automated review by oss-sentinel-ai*
